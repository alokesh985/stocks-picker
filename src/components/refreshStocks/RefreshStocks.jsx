import React, { useState, useEffect, useCallback } from "react";

import PropTypes from 'prop-types';

// Lodash
import _noop from 'lodash/noop';

// Utils
import toaster from "react-hot-toast";

// Constants
import { DEFAULT_REFRESH_INTERVAL } from "./refreshStocks.constants";
import { KEYS } from "../autoComplete/autoComplete.constants";
import { EMPTY_ARRAY } from "../../constants/general.constants";

// Helpers
import {
  convertMilliSecondsToSeconds,
  convertSecondsToMilliSeconds,
  fetchAllStockData,
} from "./refreshStocks.helpers";
import { getApiData } from "../../helpers/helpers.general";

// Styles
import styles from "./refreshStocks.module.scss";

const RefreshStocks = ({
  setSelectedStocks,
  selectedStocks,
  setStockOverviewLoading,
}) => {
  const [refreshInterval, setRefreshInterval] = useState(
    convertSecondsToMilliSeconds(DEFAULT_REFRESH_INTERVAL)
  );

  const formatAndSetUpdatedStocks = useCallback(
    (responses) => {
      const updatedStocks = responses.map(response => {
        const [overviewDataResponse, graphDataResponse] = response;
        const stockOverview = getApiData(overviewDataResponse);
        const dailyTimeSeries = getApiData(graphDataResponse);
        return { stockOverview, dailyTimeSeries }
      });
      setSelectedStocks(updatedStocks)
    },
    [setSelectedStocks]
  );

  // Fetching stock overview in some interval
  useEffect(() => {
    const interval = setInterval(() => {
      setStockOverviewLoading(true);
      const fetchStockDataPromise = fetchAllStockData(selectedStocks);
      fetchStockDataPromise
        .then(formatAndSetUpdatedStocks)
        .catch(() => toaster.error("Failed to refresh stocks"))
        .finally(() => setStockOverviewLoading(false));
    }, refreshInterval);
    return () => clearInterval(interval);
  }, [
    selectedStocks,
    refreshInterval,
    setSelectedStocks,
    setStockOverviewLoading,
    formatAndSetUpdatedStocks,
  ]);

  const handleSetNewInteval = (e) => {
    if (e.keyCode === KEYS.ENTER) {
      setRefreshInterval(convertSecondsToMilliSeconds(e.target.value));
    }
  };

  return (
    <div className={styles.refreshStocksContainer}>
      <div className={styles.refreshInputContainer}>
        <div className={styles.refreshAfterLabel}>
          Refresh After (seconds):{" "}
        </div>
        <input
          type="text"
          placeholder="Enter time in seconds and press Enter"
          onKeyDown={handleSetNewInteval}
        />
      </div>
      <div
        className={styles.refreshingAfterMessage}
      >{`Refreshing every ${convertMilliSecondsToSeconds(
        refreshInterval
      )} second(s)!`}</div>
    </div>
  );
};

RefreshStocks.propTypes = {
  setSelectedStocks: PropTypes.func,
  selectedStocks: PropTypes.array,
  setStockOverviewLoading: PropTypes.func,
};

RefreshStocks.defaultProps = {
  setSelectedStocks: _noop,
  selectedStocks: EMPTY_ARRAY,
  setStockOverviewLoading: _noop,
};

export default RefreshStocks;
