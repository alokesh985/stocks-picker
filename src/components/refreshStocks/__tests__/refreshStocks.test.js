// Utils
import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";

// Components
import RefreshStocks from "../RefreshStocks";

// Helpers
import {
  convertMilliSecondsToSeconds,
  convertSecondsToMilliSeconds,
} from "../refreshStocks.helpers";

describe("Testing the RefreshStocks component", () => {
  test("Should match snapshot", () => {
    const tree = renderer.create(<RefreshStocks />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Should handle input change", () => {
    render(<RefreshStocks />);
    const searchInput = screen.getByPlaceholderText(
      "Enter time in seconds and press Enter"
    );
    fireEvent.change(searchInput, { target: { value: "Intuit" } });
    expect(searchInput.value).toBe("Intuit");
  });
});

describe("Testing the RefreshStocks helpers", () => {
  test("Should convert milliseconds to seconds", () => {
    const result = convertMilliSecondsToSeconds(1000);
    expect(result).toBe(1);
  });

  test("Should convert seconds to milliseconds", () => {
    const result = convertSecondsToMilliSeconds(1);
    expect(result).toBe(1000);
  });
});
