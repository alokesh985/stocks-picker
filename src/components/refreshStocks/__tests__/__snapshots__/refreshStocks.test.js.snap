// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the RefreshStocks component Should match snapshot 1`] = `
<div
  className="refreshStocksContainer"
>
  <div
    className="refreshInputContainer"
  >
    <div
      className="refreshAfterLabel"
    >
      Refresh After (seconds):
       
    </div>
    <input
      onKeyDown={[Function]}
      placeholder="Enter time in seconds and press Enter"
      type="text"
    />
  </div>
  <div
    className="refreshingAfterMessage"
  >
    Refreshing every 60 second(s)!
  </div>
</div>
`;
